### ISSUE TITLE
Document what .cc means

### ISSUE THREAD

**Issue by nh2** [NixOS/nixpkgs#20270]
Hi,

I have a Haskell project I'm building with Stack and the corresponding nix helper `haskell.lib.buildStackProject` (similar to [this one here](https://github.com/tweag/sparkle/blob/17c04630642c4266c228a604e07a8e814961d620/shell.nix#L13).

My project doesn't build unless in the `buildInputs` I specify `gcc5`; if not, it complains that that things like `m` (from `libm`) are missing. So far so good.

My project also doesn't build unless I add `gcc5.cc`.

Where is this `.cc` attribute declared? I cannot find it in [`nixpkgs/pkgs/development/compilers/gcc/5/default.nix `](https://github.com/NixOS/nixpkgs/blob/5de2857a54b74907b4817203dc31f2deb12a9d45/pkgs/development/compilers/gcc/5/default.nix).

Also, what does `.cc` mean? "C compiler"? If so, why does adding `gcc5.cc` ensure that my binary links correctly against `libstdc++.so`? Does `cc` stand for C++ (as in the `.cc` file extension) after all?

Finally, my project also doesn't build unless I add `gfortran.cc`. Now what does `.cc` mean in this case?

I've checked the Nix, Nixos, and Nixpkgs manuals and couldn't find any reference to `.cc`.

It would be nice if you could answer these questions for me, and maybe we can add some appropriate documentation so that it's clear in general. Thank you!

**Comment by edolstra** [comment]
`.cc` means a C/C++ compiler with a command-line syntax compatible with GCC, i.e. GCC or Clang. In fact it was called `.gcc` until recently.

In general it should not be necessary to specify `gcc` as a build input, but maybe `buildStackProject` interferes with this in some way. (@peti?) Also note that `libm` is part of Glibc, not GCC.

**Comment by peti** [comment]
`libm` is provided by *glibc*, not gcc5. I see no reason why it should be necessary to add the compiler to the `buildInputs`. In fact, chances are that this addition interferes with the build rather than fix it. (I'm speculating, though; I haven't actually tested it. I don't use `buildStackProject` myself.)

**Comment by nh2** [comment]
> .cc means a C/C++ compiler with a command-line syntax compatible with GCC

@edolstra you mean "C/C++/Fortran compiler", or really "any compiler with command-line syntax like GCC? Otherwise `gfortran.cc` would not make sense, would it?

> I see no reason why it should be necessary to add the compiler to the buildInputs

@peti there might be other things at work, e.g. that I didn't depend on glibc but adding `gcc5` somehow exposed it, e.g. via `propagatedBuildInputs` or similar? I'm not sure.

Does it make sense that I should depend on `gcc5.cc` to get `libstdc++.so`, or is there a better way to achieve that?

**Comment by vcunat** [comment]
There's possibly a difference in whether the stdc++ library gets into `-L` directives, but gcc should just find it fine via different ones that are passed. (Some weird project might still rely on something else.)

**Comment by vcunat** [comment]
For now I believe we assume primarily that it contains `cc` and `c++` executables with all around, and accepts the basic gcc-like flags (we use gcc and clang in there).

**Comment by edolstra** [comment]
You cannot count on stdenv providing a Fortran compiler (and indeed it doesn't, AFAIK). However `gfortran` happens to use the same gcc/gcc-wrapper Nix expressions so `gfortran.cc` does mean the unwrapped compiler package.

Actually, the question of what `.cc` means is a bit ambiguous because that name is overloaded. `stdenv.cc` is the wrapped C/C++ compiler of stdenv, while `stdenv.cc.cc` is the unwrapped compiler. It probably would have been better to use different names, but that ship has sailed...

**Comment by nh2** [comment]
> It probably would have been better to use different names, but that ship has sailed...

As long as it's documented somehow, having this kind of legacy seems OK, but without this explanation it can be pretty rough for a beginner to figure out what it means.

**Comment by Profpatsch** [comment]
(triage) is there still something to be done here?

**Comment by nh2** [comment]
Yes, I think we want to put the above expert commentary into explicit documentation.

**Comment by Profpatsch** [comment]
@nh2 are you interested in preparing a pull request?

**Comment by nh2** [comment]
@Profpatsch I'd be more comfortable if this were written up by somebody actually in the know vs me scaping things together from the above small windows into people's minds.

**Comment by Profpatsch** [comment]
```
nix-repl> gcc.cc.meta.name
"gcc-7.3.0"

nix-repl> gcc.meta.name
"gcc-7.3.0"

nix-repl> stdenv.cc.meta.name
"gcc-7.3.0"

nix-repl> stdenv.cc.meta.description
"GNU Compiler Collection, version 7.3.0 (wrapper script)"
```

I’m actually not sure what should be documented in this case. The `cc` attribute is more of a Unix convention for a C compiler with a semi-conventional interface and nixpkgs mirrors that in some places, e.g. `stdenv` has it, the `gcc` one is just `gcc` itself, so I’m not sure how useful that is.

**Comment by oxij** [comment]
I agree. "CC" env var is the standard Makefile thing that points to the "C Compiler" executable for this system. Seems pretty obvious to me.

**Comment by nh2** [comment]
> I agree. "CC" env var is the standard Makefile thing that points to the "C Compiler" executable for this system. Seems pretty obvious to me.

That part is clear, please see the issue description for what the confusing bits are:

> Project doesn't build when `gcc5` is added to `build-depends` but `gcc5.cc` isn't

> why does adding gcc5.cc ensure that my binary links correctly against libstdc++.so?

> `gfortran.cc`

None of these things make sense from the classic Makefile CC env var perspective. Further, clasically `CC` describes an _executable_ (e.g. `CC=/usr/bin/cc`).

And as @edolstra said above,

> `gfortran.cc` does mean the unwrapped compiler package

which is definitely not something that's explained by the Makefile convention of calling the C compiler executable `CC`.

So in summary, `nixpkgs` clearly has some concrete ideas and naming conventions. Those should be documented.

**Comment by Profpatsch** [comment]
> gfortran.cc does mean the unwrapped compiler package

Since e.g. `ghc` uses `ghc.compiler` it rather looks to me like there *aren’t* any global conventions that could be documented.

**Comment by nh2** [comment]
> there aren’t any global conventions that could be documented

I'm not suggesting that there are _global_ conventions, and I'm not sure how Haskell comes into play. It is fine to already document local conventions (as parts of the manual, like `User’s Guide to the Haskell Infrastructure`, already try to do).

The idea is that somebody building packages needs to be able to look up in documentation when to depend on `gcc5`, when to depend on `gcc5.cc`, what `.cc` means in that case and so on.

Right now building packages as somebody who hasn't read all nixpkgs source code is a lot of guesswork; trying out random things and digging deep into the code is not what I want to do as a user. That is why I'm asking for these things to be documented.

**Comment by Profpatsch** [comment]
@nh2 Since you have gained a bit of insight in this matter, would you be willing to update the documentation now? You can even put it in the place where you’d have searched for it, so it’s a double win. :)

**Comment by tomberek** [comment]
or close?
